src/action/timeline.rs

```diff
@@ -0,0 +1,168 @@
+        path: Some(Vec::from(p.to_string_lossy().as_bytes())),
+        mode: Some(md.mode()),
+        size: Some(md.size()),
+        dev: Some(md.dev()),
+        ino: Some(md.dev()),
+        uid: Some(md.uid() as i64),
+        gid: Some(md.gid() as i64),
+        atime_ns: Some(md.atime_nsec() as u64),
+        ctime_ns: Some(md.ctime_nsec() as u64),
+        mtime_ns: Some(md.mtime_nsec() as u64)
+    }
+}
+
+impl RecurseState {
+    /// Recursively traverses path specified as root, sends gzchunked stat data to session.
+    fn recurse<S: Session>(&mut self, root: &PathBuf, session: &mut S) -> session::Result<()> {
+        let statentry = match fs::symlink_metadata(&root) {
+            Ok(v) => v,
+            Err(_) => return Ok(())
+        };
+        let statproto = entry_from_stat(&statentry, root);
+        let mut statdata: Vec<u8> = Vec::new();
+        prost::Message::encode(&statproto, &mut statdata)?;
+        self.encoder.write_all(&(statdata.len() as u64).to_be_bytes())?;
+        self.encoder.write_all(statdata.as_slice())?;
+        self.encoder.flush()?;
+        if self.encoder.get_ref().len() >= BLOCK_SIZE {
+            self.flush(session)?;
+        }
+        if statentry.is_dir() && statentry.dev() == self.device {
+            for entry in match fs::read_dir(root) {
+                Ok(v) => v,
+                Err(_) => return Ok(())
+            } {
+                self.recurse(&entry?.path(), session)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Sends currently accumulated gzchunked data to transfer store.
+    fn flush<S: Session>(&mut self, session: &mut S) -> session::Result<()> {
+        self.encoder.try_finish()?;
+        let data = self.encoder.get_ref().clone();
+        self.ids.push(Vec::from(Sha256::digest(data.as_slice()).as_slice()));
+        session.send(session::Sink::TRANSFER_STORE, ChunkResponse{ data })?;
+        // TODO(xtsm) add session.progress here
+        self.encoder = GzEncoder::new(Vec::new(), Compression::default());
+        Ok(())
+    }
+}
+
+/// Handles requests for the timeline action.
+pub fn handle<S: Session>(sess: &mut S, request: Request) -> session::Result<()> {
+
+    let mut rs = RecurseState {
+        device: match fs::symlink_metadata(&request.root) {
+            Ok(v) => v.dev(),
+            Err(_) => {
+                sess.reply(Response {ids : Vec::new()})?;
+                return Ok(());
+            }
+        },
+        ids: Vec::new(),
+        encoder: GzEncoder::new(Vec::new(), flate2::Compression::default()),
+    };
+    rs.recurse(&request.root, sess)?;
+    rs.flush(sess)?;
+    sess.reply(Response {ids: rs.ids})?;
+
+    Ok(())
+}
+
+impl super::Request for Request {
+
+    type Proto = rrg_proto::TimelineArgs;
+
+    fn from_proto(proto: rrg_proto::TimelineArgs) -> Request {
+        // TODO fix unwrap
+        Request {
+            root: PathBuf::from(String::from_utf8(proto.root.unwrap()).unwrap())
```

Comment:
What if the file system does not use UTF-8 encoding (which is very often the case)? For this very reason the proto message uses `bytes` for representing paths, rather than `string` (which requires to be a valid UTF-8).